state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+        })?;
+
+        // Fetch the new epoch from storage
+        let epoch = AccountState::try_from(&account_state_blob)
+            .and_then(|state| {
+                Ok(state
+                    .get_configuration_resource()?
+                    .ok_or_else(|| {
+                        Error::UnexpectedErrorEncountered(
+                            "Configuration resource does not exist!".into(),
+                        )
+                    })?
+                    .epoch())
+            })
+            .map_err(|error| {
+                Error::UnexpectedErrorEncountered(format!(
+                    "Failed to fetch configuration resource! Error: {:?}",
+                    error
+                ))
+            })?;
+
+        // Return the config payload
+        Ok(OnChainConfigPayload::new(
+            epoch,
+            Arc::new(
+                ON_CHAIN_CONFIG_REGISTRY
+                    .iter()
+                    .cloned()
+                    .zip_eq(on_chain_configs)
+                    .collect(),
+            ),
+        ))
+    }
+}
+
+#[async_trait]
+impl EventNotificationSender for EventSubscriptionService {
+    async fn notify_events(
+        &mut self,
+        version: Version,
+        events: Vec<ContractEvent>,
+    ) -> Result<(), Error> {
+        if events.is_empty() {
+            return Ok(()); // No events!
+        }
+
+        // Notify event subscribers and check if a reconfiguration event was processed
+        let reconfig_event_processed = self.notify_event_subscribers(version, events)?;
+
+        // If a reconfiguration event was found, also notify the reconfig subscribers
+        // of the new configuration values.
+        if reconfig_event_processed {
+            self.force_notify_reconfigurations(version).await
+        } else {
+            Ok(())
+        }
+    }
+
+    async fn force_notify_reconfigurations(&mut self, version: Version) -> Result<(), Error> {
+        if self.reconfig_subscriptions.is_empty() {
+            return Ok(()); // No reconfiguration subscribers!
+        }
+
+        let new_configs = self.read_on_chain_configs(version)?;
+        for (_, reconfig_subscription) in self.reconfig_subscriptions.iter_mut() {
+            reconfig_subscription.notify_subscriber_of_configs(version, new_configs.clone())?;
+        }
+
+        Ok(())
+    }
+}
+
+/// A unique ID used to identify each subscription.
+type SubscriptionId = u64;
+
+/// A single event subscription, holding the subscription identifier, channel to
+/// send the corresponding notifications and a buffer to hold pending events.
+#[derive(Debug)]
+struct EventSubscription {
+    pub subscription_id: SubscriptionId,
+    pub event_buffer: Vec<ContractEvent>,
+    pub notification_sender: channel::diem_channel::Sender<(), EventNotification>,
+}
+
+impl EventSubscription {
+    fn buffer_event(&mut self, event: ContractEvent) {
+        self.event_buffer.push(event)
+    }
```

Comment:
Add event?

So, why do we have a buffer in addition to the sender channel?  Why not just use the channel as the buffer?  Is this just to collect them all into a group?

Reply:
I went for `buffer` over `add` to explicitly call out that the event wasn't being sent across the wire yet, but buffered internally first. 🤔 The reasons for this buffering approach are: (i) it allows us to send a single notification for each blockchain version (i.e., grouping events by version); and (ii) it might even be faster (but this is speculation and not something super important right now 😆).

We could send everything one at a time, and provide versions for each message (kinda like a stream), but I figure this is simpler for now.